### ISSUE TITLE
http server benchmark.

### ISSUE THREAD

**Issue by patrickkh7788** [nanomsg/nng#727]
build with today master,  Release version and O3.  return the default 404 page.
wrk report 18K req/sec.

Test with simple crystal http server it is 96.85k(not 404 page,  not sure how many thread, from the document it seems is 1).

 maybe we can provide benchmark on the readme page.

**Comment by gdamore** [comment]
The HTTP server framework is not highly optimized for performance -- frankly I don't expect it to be extraordinarily fast compared to other implementations.  What test suite did you use?

I'd also like to know how you configured the server?  You'll get a 404 if you just start a server with no extra files configured -- we don't come with any sort of builtin file for "/", because being a general purpose web server isn't part of main mission here. :-)

**Comment by patrickkh7788** [comment]
I use `wrk -t1 -c200 -d1s http://127.0.0.1:8081` to test,  and expect it return 404.

I think maybe there is memory alloc and free for each request to slow it down?

If there a way to avoid alloc in the process ?

**Comment by gdamore** [comment]
Yes, there are probably numerous allocations, and other areas that may not be ideally efficient.  Part of the problem is that we are trying to build a general purpose HTTP server for building application services -- including an asynchronous framework to allow lots of background processing (it really started out being something that could handle initiating websocket stuff).  It was not designed for extreme performance.

Having said that, I'm sure that some analysis could reveal opportunities to optimize this code.  It's still fairly young, and I've been more focused on correctness and flexibility.  Given that, I'm fairly happy with 18k req/sec.

If you want to deep dive and suggest specific changes, I'm open to hearing about them.  Elimination the dynamic allocations isn't going to be something easy to do, at least not for now.

**Comment by zhangxinlong633** [comment]
impove or not ?

**Comment by gdamore** [comment]
18K / sec seems low. My current tests give about 100K ... and I'm looking at improving that to > 200K (this on a MBP with an M4.) Admittedly tests run in 2020 on older hardware likely different than in 2026 with newer hardware.

**Comment by gdamore** [comment]
There were a number of optimizations done in the past few years so the performance is indeed much higher.  Adding more task threads helps enormously if this is something you want (its a compile time tunable).  I'm adding a little more optimization then going to close this out.
